### PR TITLE
De-underscore @frozen, apply it to structs

### PR DESCRIPTION
**Pull request by airspeedswift**
Implementation for [SE-260](https://github.com/apple/swift-evolution/blob/master/proposals/0260-library-evolution.md)

This leaves classes and vars with `@_fixed_layout` for now. It also seems like leaving the compiler using the term "fixed layout" internally makes sense.

I've added some tests for the old names, but migrated most of the tests to the new. Not sure if more tests of the old names are merited.

IHaveNoIdeaWhatImDoing.gif, sorry if this doesn't match some conventions.

### CODE REVIEW

**Comment by airspeedswift** [comment]
Oops, can you make a PR a draft retroactively?

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -1354,8 +1354,8 @@ WARNING(pattern_type_not_usable_from_inline_warn,none,
         "%select{%select{variable|constant}0|property}1 "
         "should be '@usableFromInline' or public",
         (bool, bool))
-ERROR(pattern_type_not_usable_from_inline_fixed_layout,none,
-      "type referenced from a stored property in a '@_fixed_layout' struct must "
+ERROR(pattern_type_not_usable_from_inline_frozen,none,
```
Comment (slavapestov) [inline include/swift/AST/DiagnosticsSema.def]: Do we use this diagnostic for classes too? If so it needs a %choice

**Review by slavapestov — APPROVED** [review]

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (jrose-apple) [reply include/swift/AST/DiagnosticsSema.def]: I bet we do, but it also says "struct" in the message, so…

Comment (airspeedswift) [reply include/swift/AST/DiagnosticsSema.def]: Yeah it was always struct-specific even if it applies to both. I'll happily correct it but I'd like to do that in a follow-on PR than lump it in here.

File: lib/Sema/TypeCheckAttr.cpp [hunk 2]
```diff
@@ -1932,6 +1932,11 @@ void AttributeChecker::visitSpecializeAttr(SpecializeAttr *attr) {
 }
 
 void AttributeChecker::visitFixedLayoutAttr(FixedLayoutAttr *attr) {
+  if (isa<StructDecl>(D)) {
+    TC.diagnose(attr->getLocation(), diag::fixed_layout_struct)
+      .fixItReplace(attr->getRange(), "@frozen");
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckAttr.cpp]: I think this needs to be `getRangeWithAt` (or that you should omit the `@` from your replacement).

File: lib/Sema/TypeCheckAttr.cpp [hunk 3]
```diff
@@ -2442,16 +2447,25 @@ void AttributeChecker::visitImplementsAttr(ImplementsAttr *attr) {
...
+    if (!ED->getModuleContext()->isResilient()) {
+      diagnoseAndRemoveAttr(attr, diag::enum_frozen_nonresilient, attr);
+      return;
+    }
 
-  if (!ED->getModuleContext()->isResilient()) {
-    diagnoseAndRemoveAttr(attr, diag::enum_frozen_nonresilient, attr);
-    return;
+    if (ED->getFormalAccess() < AccessLevel::Public &&
+        !ED->getAttrs().hasAttribute<UsableFromInlineAttr>()) {
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckAttr.cpp]: Do you still need this check if you're going to do the one below?

File: lib/Sema/TypeCheckAttr.cpp [hunk 4]
```diff
@@ -2442,16 +2447,25 @@ void AttributeChecker::visitImplementsAttr(ImplementsAttr *attr) {
 }
 
 void AttributeChecker::visitFrozenAttr(FrozenAttr *attr) {
-  auto *ED = cast<EnumDecl>(D);
+  if (auto *ED = dyn_cast<EnumDecl>(D)) {
+    if (!ED->getModuleContext()->isResilient()) {
+      diagnoseAndRemoveAttr(attr, diag::enum_frozen_nonresilient, attr);
```
Comment (jrose-apple) [inline lib/Sema/TypeCheckAttr.cpp]: This check probably applies to structs as well.

File: test/decl/enum/frozen.swift [hunk 5]
```diff
@@ -10,4 +10,5 @@ internal enum Outer {
 
 @frozen @usableFromInline enum NotPublicButVersioned {} // no-warning
 
-@_frozen enum DeprecationWarning {} // expected-warning {{@frozen has no effect on non-public enums}} {{1-9=}}
```
Comment (jrose-apple) [inline test/decl/enum/frozen.swift]: You lost the warning here?

**Review by jrose-apple — COMMENTED** [review]
Looks pretty good. Thanks, Ben!

**Comment by airspeedswift** [comment]
@swift-ci please test

File: lib/Sema/TypeCheckAttr.cpp (hunk 2 above)
Comment (airspeedswift) [reply lib/Sema/TypeCheckAttr.cpp]: I think I'm missing something, this looks right as-is to me:
```
'@frozen' attribute is now used for fixed-layout structs
@_fixed_layout
^~~~~~~~~~~~~~
@frozen
```

File: lib/Sema/TypeCheckAttr.cpp (hunk 3 above)
Comment (airspeedswift) [reply lib/Sema/TypeCheckAttr.cpp]: This one's just a warning for enums, whereas the struct one's an error. Though I'm not sure if that inconsistency is desirable.

File: test/decl/enum/frozen.swift (hunk 5 above)
Comment (airspeedswift) [reply test/decl/enum/frozen.swift]: not sure why I removed that test, back now.

File: lib/Sema/TypeCheckAttr.cpp (hunk 3 above)
Comment (airspeedswift) [reply lib/Sema/TypeCheckAttr.cpp]: (this was the old behavior too fwiw)

File: lib/Sema/TypeCheckAttr.cpp (hunk 2 above)
Comment (jrose-apple) [reply lib/Sema/TypeCheckAttr.cpp]: Are you sure you weren't seeing

```
'@frozen' attribute is now used for fixed-layout structs
@_fixed_layout
^~~~~~~~~~~~~~
 @frozen
```

? Which is a different thing.

File: lib/Sema/TypeCheckAttr.cpp (hunk 4 above)
Comment (jrose-apple) [reply lib/Sema/TypeCheckAttr.cpp]: Still not resolved.

File: include/swift/AST/DiagnosticsSema.def [hunk 6]
```diff
@@ -4136,6 +4136,15 @@ ERROR(fixed_layout_attr_on_internal_type,
       "%select{private|fileprivate|internal|%error|%error}1",
       (DeclName, AccessLevel))
 
+WARNING(fixed_layout_struct,
+      none, "'@frozen' attribute is now used for fixed-layout structs", ())
```
Comment (jrose-apple) [inline include/swift/AST/DiagnosticsSema.def]: Nitpick: please pick either `'@frozen'` or `'frozen' attribute` (both here and below); having both is redundant.

File: lib/Sema/TypeCheckAttr.cpp (hunk 4 above)
Comment (airspeedswift) [reply lib/Sema/TypeCheckAttr.cpp]: Similar thing, this is the existing behavior, so I'd suggest landing and then a follow-up PR can make things more consistent across the two.

File: lib/Sema/TypeCheckAttr.cpp (hunk 2 above)
Comment (airspeedswift) [reply lib/Sema/TypeCheckAttr.cpp]: A `@_fixed_layout public struct FixedDeprecationWarning {} // expected-warning {{'@frozen' attribute is now used for fixed-layout structs}} {{1-15=@frozen}}` test passes.

**Pull request merged**